### ISSUE TITLE
1.78: initially "mount" is very slow [not-fixed]

### ISSUE THREAD

**Issue by onlyjob** [chrislusf/seaweedfs#1322]
After `weed mount -dir=/mnt/weed`, commands like `df -h /mnt/weed`, `ls -l /mnt/weed` initially take _minutes_(!) to return anything (i.e. long time to first result).

I'd expect it to be faster to list content of the mount directory. Full walk of directory tree is not required for this.

The unfortunate side effect of this problem is that applications freeze for a while on first attempt to access the mount directory shortly after it has been mounted.

Perhaps `weed mount` should not return so fast if mount is not ready yet?

Anyway it would be best to optimise initial access to the directory.

**Comment by chrislusf** [comment]
How large is the filer folder? Maybe mount a specific filer folder instead of the filer root.

Disabling the metadata caching may help, but not really recommended.
`weed mount -asyncMetaDataCaching=false`

The problem is that the `weed mount` does not have a way to declare "I am ready to accept requests". I can add some printouts when it is ready.

**Comment by chrislusf** [comment]
Added fix https://github.com/chrislusf/seaweedfs/commit/ac48c899e8ccf959940b11033c2222796519386f

**Comment by onlyjob** [comment]
The folder that I'm trying to list have just one file and one directory. That directory have few sub-directories that have several directories with many files. Total number of files is around 30_000_000 (30 million).

`asyncMetaDataCaching=false` did not help. Mounting sub-directory is not an option as all the files are required for application.

This problem is in conflict with two project objectives ("store billions of files", "serve files fast").

Growing the file system will increase already unmanageable mount delay to prohibitively unbearable.

Printout is _not_ a fix. A better solution is needed.

Yet another issue closed too fast in unsatisfactory manner.  
Such issue management goes against any best practice for handling issues.  
Please reopen.

**Comment by chrislusf** [comment]
Please help to design with a better design.

**Comment by chrislusf** [comment]
A side question: the `weed mount` would usually only be started once for an OS, unless restarts often. What is the use case you are trying? It is possible to resume from last metadata sync checkpoint to reduce restart time.

**Comment by onlyjob** [comment]
Slow restart is problematic on its own (e.g. long _fsck_ time). On this instance file system _appears_ to be ready (since `mount` command succeeded) but in fact FUSE-mount remains unusable for a while, causing any application that touches the mount point to freeze.

This is a situation when long initialisation time creates cascading problems in services depending of data from FUSE mount (e.g. timeouting services).

**Comment by chrislusf** [comment]
The fix in 1.78 release already has the fix to actually mount the directory after the meta data is in sync.

It does not have anything to prevent other programs to read the directory when it is synchronizing data in progress before the actual mounting.

**Comment by onlyjob** [comment]
Here is a little reproducer:

```
# mount -v /mnt/weed; sleep 2; time sh -c 'ls -l /mnt/weed | wc -l'

This is SeaweedFS version 8000GB 1.78 linux amd64
mount point owner uid=0 gid=0 mode=drwxr-xr-x
current uid=0 gid=0

4

real    9m7.465s
user    0m0.005s
sys     0m0.001s
```
Nine(!) minutes to list directory with 4 entries.

Changing `-cacheCapacityMB` does not seem to make any difference (I tried values `0`, `200` and (default) `1000`). `/tmp` is _tmpfs_-mounted.

**Comment by chrislusf** [comment]
There has been some build issue, not sure whether your build included https://github.com/chrislusf/seaweedfs/commit/ac48c899e8ccf959940b11033c2222796519386f

With the above commit, the behavior is, 
1. synchronize the metadata, which can take 9 minutes in your case.
2. mount the folder.

In the first step, if mounted to an empty folder, the `ls -l` will work, but return empty. Only after the second step, the `ls -l` can return the actual items on the filer.

**Comment by chrislusf** [comment]
One possible way is to provide a `weed mount.init` command to synchronize the metadata, and `weed mount` just to mount the folder. Seems too complicated though.

**Comment by onlyjob** [comment]
Another side effect of this problem:

```
# fusermount -u /mnt/weed
fusermount: failed to unmount /mnt/weed: Device or resource busy
```

Basically I can't _umount_ until initialisation is completed. Once mounted, I have to wait for 9 minutes before it would be possible to umount...

`weed mount.init` is probably not a good idea...

**Comment by chrislusf** [comment]
Current code v1.79 should work for you then. The actual mount does not happen until the metadata sync.

**Comment by onlyjob** [comment]
Hmm, I'll test 1.79 then. Thanks.

**Comment by chrislusf** [comment]
In 1.82, the mount is changed to lazy meta data synchronization.
